File: frontend/src/types/emergency.ts
import type { QueryCenter } from './spatial'
import type { IsochroneResponse, NetworkSnap } from './routing'

export type IncidentType = 'medical' | 'fire'
export type EmergencyStatus =
  | 'idle'
  | 'selecting'
  | 'loading'
  | 'success'
  | 'snap_failed'
  | 'no_facilities'
  | 'error'

export interface EmergencyFacilityCandidate {
  network_rank: number
  poi_id: number
  osm_type: string
  osm_id: string
  name: string | null
  category: 'healthcare' | 'emergency'
  subcategory: 'hospital' | 'fire_station'
  lon: number
  lat: number
  node_id: number | null
  edge_id: number | null
  fraction: number | null
  snapped_lon: number | null
  snapped_lat: number | null
  facility_snap_distance_m: number
  straight_distance_m: number
  response_time_s: number
  response_time_min: number
}

export interface EmergencyResponse {
  incident_type: IncidentType
  incident: QueryCenter
  incident_snap: NetworkSnap
  facility_statistics: {
    total: number
    mapped: number
    unmapped: number
    reachable: number
    unreachable: number
    returned: number
  }
  recommended_facility: EmergencyFacilityCandidate
  candidate_facilities: EmergencyFacilityCandidate[]
  response_route: {
    type: 'Feature'
    geometry: { type: 'LineString'; coordinates: number[][] }
    properties: {
      network_distance_m: number
      response_time_s: number
      response_time_min: number
      edge_count: number
      edge_ids: number[]
      routing_direction: 'facility_to_incident'
    }
  }
  response_isochrones: IsochroneResponse['isochrones']
  comparison: {
    straight_nearest_facility_id: number
    network_best_facility_id: number
    straight_nearest_is_network_best: boolean
  }
  model_metadata: {
    cost_model: 'static_travel_time'
    directed: true
    routing_direction: 'facility_to_incident'
    service_area_method: 'reachable_vertex_concave_hull'
  }
}

File: backend/app/schemas/emergency.py
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.routing import IsochroneFeatureCollection, NetworkSnap, RoutePoint

IncidentType = Literal["medical", "fire"]


class EmergencyResponseRequest(BaseModel):
    incident: RoutePoint
    incident_type: IncidentType
    max_snap_m: int = Field(default=500, ge=100, le=1000)
    candidate_limit: int = Field(default=3, ge=1, le=10)


class EmergencyFacilityCandidate(BaseModel):
    network_rank: int
    poi_id: int
    osm_type: str
    osm_id: str
    name: str | None
    category: Literal["healthcare", "emergency"]
    subcategory: Literal["hospital", "fire_station"]
    lon: float
    lat: float
    node_id: int | None
    edge_id: int | None = None
    fraction: float | None = Field(default=None, ge=0, le=1)
    snapped_lon: float | None = None
    snapped_lat: float | None = None
    facility_snap_distance_m: float
    straight_distance_m: float
    response_time_s: float
    response_time_min: float


class EmergencyFacilityStatistics(BaseModel):
    total: int
    mapped: int
    unmapped: int
    reachable: int
    unreachable: int
    returned: int


class EmergencyRouteProperties(BaseModel):
    network_distance_m: float
    response_time_s: float
    response_time_min: float
    edge_count: int
    edge_ids: list[int]
    routing_direction: Literal["facility_to_incident"] = "facility_to_incident"


class EmergencyRouteFeature(BaseModel):
    type: Literal["Feature"] = "Feature"
    geometry: dict[str, Any]
    properties: EmergencyRouteProperties


class EmergencyComparison(BaseModel):
    straight_nearest_facility_id: int
    network_best_facility_id: int
    straight_nearest_is_network_best: bool


class EmergencyModelMetadata(BaseModel):
    cost_model: Literal["static_travel_time"] = "static_travel_time"
    directed: Literal[True] = True
    routing_direction: Literal["facility_to_incident"] = "facility_to_incident"
    service_area_method: Literal["reachable_vertex_concave_hull"] = (
        "reachable_vertex_concave_hull"
    )


class EmergencyResponse(BaseModel):
    incident_type: IncidentType
    incident: RoutePoint
    incident_snap: NetworkSnap
    facility_statistics: EmergencyFacilityStatistics
    recommended_facility: EmergencyFacilityCandidate
    candidate_facilities: list[EmergencyFacilityCandidate]
    response_route: EmergencyRouteFeature
    response_isochrones: IsochroneFeatureCollection
    comparison: EmergencyComparison
    model_metadata: EmergencyModelMetadata
